tools/tpm2_clearlock.c

```diff
@@ -0,0 +1,125 @@
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of Intel Corporation nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdbool.h>
+#include <stdlib.h>
+
+#include "log.h"
+#include "tpm2_password_util.h"
+#include "tpm2_tool.h"
+#include "tpm2_util.h"
+
+typedef struct clearlock_ctx clearlock_ctx;
+struct clearlock_ctx {
+    bool clear;
+    bool platform;
+    bool password;
+    TPMS_AUTH_COMMAND session_data;
+};
+
+static clearlock_ctx ctx = {
+    .clear = false,
+    .platform = false,
+    .password = false,
+    .session_data = TPMS_AUTH_COMMAND_INIT(TPM2_RS_PW),
+};
+
+static bool clearlock(TSS2_SYS_CONTEXT *sapi_context) {
+
+    TSS2L_SYS_AUTH_COMMAND sessionsData = { 1, { ctx.session_data }};
+    TSS2L_SYS_AUTH_RESPONSE sessionsDataOut;
+    TPMI_RH_CLEAR rh = TPM2_RH_LOCKOUT;
+    TPMI_YES_NO disable = YES;
+
+    LOG_INFO ("Sending TPM2_ClearControl(%s) command on %s",
+            ctx.clear ? "CLEAR" : "SET",
+            ctx.platform ? "TPM2_RH_PLATFORM" : "TPM2_RH_LOCKOUT");
+
+    if (ctx.platform)
+        rh = TPM2_RH_PLATFORM;
+
+    if (ctx.clear)
+        disable = NO;
+
+    TSS2_RC rc = TSS2_RETRY_EXP(Tss2_Sys_ClearControl (sapi_context,
+            rh, &sessionsData, disable, &sessionsDataOut));
+
+    if (rc != TPM2_RC_SUCCESS && rc != TPM2_RC_INITIALIZE) {
+        LOG_ERR ("Tss2_Sys_ClearControl failed: 0x%x", rc);
+        return false;
+    }
+
+    LOG_INFO ("Success. TSS2_RC: 0x%x", rc);
+    return true;
+}
+
+static bool on_option(char key, char *value) {
+
+    bool result;
+
+    switch (key) {
+    case 'c':
+        ctx.clear = true;
+        break;
+    case 'p':
+        ctx.platform = true;
+        break;
+    case 'L':
+        result = tpm2_password_util_from_optarg(value, &ctx.session_data.hmac);
+        if (!result) {
+            LOG_ERR("Invalid lockout password, got\"%s\"", value);
+            return false;
+        }
+        ctx.password = true;
```

Comment:
set but not used?

Reply:
This is a leftover of the previous code version. 

Thanks for your comments ! I'm going to correct all those little things today. 

BR, 

-- Emmanuel Deloget